### DISCUSSION TITLE
{COPY} isn't working for me and neither is {COPYFILE} or {COPYDIR}

### DISCUSSION THREAD

**Discussion by AidenJWright** [premake/premake-core#2040] [Q&A]
I've been following along with [this tutorial series](https://youtu.be/sULV3aB2qeU?t=1633) and because it is a few years old, the {COPY} command has been deprecated in the intervening time. 
The original command in the video:
```lua
postbuildcommands
{
    ("{COPY} %{cfg.buildtarget.relpath} ../bin/" .. outputdir .. "/Sandbox")
}
```
Switching `{COPY}` to `{COPYDIR}` or `{COPYFILE}` has been uneventful and I'm unsure what about this command is causing me to get the error `Error: C:/dev/Hazel/premake5.lua:46: attempt to concatenate a nil value (global 'outputdir')`.

I am currently on VS2022

The suggestions in the video comments and my googling around trying to find similar struggles have been uneventful. This is my first time using Lua so please go easy on me 😅

**Comment by fjjjjjf** [comment]
I have the same problem and after I change {COPY} into {COPYDIR} ,"/Sandbox" cannot find "Hazel.dll", Who can give me some suggestions or show me your "premake5.lua",Thanks
![image](https://user-images.githubusercontent.com/106924689/223131920-3b49bc2c-2839-473f-98d6-e2950e320209.png)

**Comment by Jarod42** [comment]
In my testing [repo](https://github.com/Jarod42/premake-sample-projects), I got [project pre-postbuild](https://github.com/Jarod42/premake-sample-projects/tree/master/projects/03-prepost-build/premake5.lua) with copy/move.

> Error: C:/dev/Hazel/premake5.lua:46: attempt to concatenate a nil value (global 'outputdir')

You forgot to set value to `outputdir`, or maybe you want `"%{prj.targetdir}"` instead

**Comment by Elscol-Loro** [comment]
I am following the same tutorial and I found that in the example given in the tutorial that the lines below do not work.

	targetdir = ("bin/" .. outputdir .. "/%{prj.name}")
	objdir = ("bin-int/" .. outputdir .. "/%{prj.name}")

but these do:

	targetdir ("bin/" .. outputdir .. "/%{prj.name}")
	objdir ("bin-int/" .. outputdir .. "/%{prj.name}")

I am still having a problem getting the DLL to copy properly, but the nil value went away.

**Reply by Jarod42** [reply]
```
targetdir = ("bin/" .. outputdir .. "/%{prj.name}")
```
is indeed wrong. It overwrites the Premake function `targetdir` by a `string`...

To copy dll, `postbuildcommands` is one option, or with recent premake,

```
files { "*.dll" }
filter "files:*.dll"
  buildaction "Copy"
filter {}
```

**Comment by huism** [comment]
I use this way: 
1. mkdir Sandbox dir
2. copy file

```
postbuildcommands
{
	("{MKDIR} %{cfg.buildtarget.directory}" .. "../Sandbox"),
	("{COPYFILE} %{cfg.buildtarget.relpath} %{cfg.buildtarget.directory}" .. "../Sandbox/")
}
```

it work on vs2022. 😃

**Comment by Jarod42** [comment]
Prefer `%[]` (See [path-in-commands](https://premake.github.io/docs/Tokens/#path-in-commands))
```
postbuildcommands
{
	"{MKDIR} %[%{!cfg.buildtarget.directory}/../Sandbox]",
	"{COPYFILE} %[%{!cfg.buildtarget.relpath}] %[%{!cfg.buildtarget.directory}/../Sandbox/]"
}
```
